ec2utils/ec2uploadimg/lib/ec2utils/ec2uploadimg.py

```diff
@@ -238,6 +242,8 @@ def _clean_up(self):
             for volume in self.created_volumes:
                 self._detach_volume(volume, True)
                 self._remove_volume(volume)
+        if self.temporary_key_created:
```

Comment:
Does this work? Just wondering because I think you will need a waiter here, otherwise it could fail with an DependencyError. Same as here https://github.com/SUSE/Enceladus/pull/179/files#diff-4e140cc0dbecb24f9927a6ce15557b99R277

Reply:
The create_key_pair call directly returns a response, which includes the created private key. So I don't think we can have a race condition here.
And for delete_key_pair I assume the same. At least the AWS docu doesn't state that this call is handled asynchronously.

But there isn't any error handling right now. I guess I should add that.

Reply:
Yes, for creating I don't see a problem but for deleting. If you attach the key to an instance, I guess you're not allowed to delete this key as long as this instance is not terminated. See LOC 240, the termination usually takes some time (a few minutes). This comment probably makes more sense here https://github.com/SUSE/Enceladus/pull/181/files#diff-4e140cc0dbecb24f9927a6ce15557b99R438

Reply:
Maybe just try to delete a key from a running instance, I bet ec2 is complaining about that ... (if not I would be surprised because there is no way to login to this machine anymore...)

Reply:
A key is not attached to an instances, what happens is that the public key is transferred into the instance. The key-pair has to exists by name when an instance is created, but it need not live until all instances using the keypair have been terminated. There should not be a problem of deleting the keypair. Anyway, by the time we are cleaning up the helper instance is on it's way out, i.e. it ie being terminated. Thus there's no problem.

Reply:
It's possible, but not recommended for the reasons you just mentioned^^ And given that the helper instance get's terminated, this should be fine.

Reply:
Thanks @rjschwei for the clarification!